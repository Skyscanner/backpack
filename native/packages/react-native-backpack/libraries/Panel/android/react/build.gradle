apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
