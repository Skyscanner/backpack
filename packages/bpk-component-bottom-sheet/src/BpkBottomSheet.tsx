/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type { SyntheticEvent, ReactNode } from "react";
import { useState } from "react";

import BpkBreakpoint, { BREAKPOINTS } from "../../bpk-component-breakpoint";
// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
import BpkCloseButton from '../../bpk-component-close-button';
import { BpkButtonLink } from '../../bpk-component-link';
import BpkNavigationBar from "../../bpk-component-navigation-bar";
import BpkText, { TEXT_STYLES } from "../../bpk-component-text/src/BpkText";
import { BpkDialogWrapper, cssModules } from "../../bpk-react-utils";

import STYLES from './BpkBottomSheet.module.scss';

const getClassName = cssModules(STYLES);

export type Props = {
  actionText?: string;
  ariaLabelledby: string;
  children: ReactNode;
  closeLabel?: string;
  closeOnEscPressed?: boolean;
  closeOnScrimClick?: boolean;
  id: string;
  onAction?: () => void;
  onClose: (
    arg0?: TouchEvent | MouseEvent | KeyboardEvent | SyntheticEvent<HTMLDialogElement, Event>,
    arg1?: {
      source: 'ESCAPE' | 'DOCUMENT_CLICK';
    },
  ) => void;
  title?: string;
  wide?: boolean;
  isOpen: boolean;
}

const BpkBottomSheet = ({
  actionText = '',
  ariaLabelledby,
  children,
  closeLabel = '',
  closeOnEscPressed = false,
  closeOnScrimClick = false,
  id,
  isOpen,
  onAction = () => null,
  onClose,
  title = '',
  wide = false
}: Props) => {
  const [exiting, setExiting] = useState(false);

  const animationTimeout = 240;

  const handleClose = (
    timeout: number,
    arg0?: TouchEvent | MouseEvent | KeyboardEvent | SyntheticEvent<HTMLDialogElement, Event>,
    arg1?: {
      source: 'ESCAPE' | 'DOCUMENT_CLICK';
    },
  ) => {
    setExiting(true)
    setTimeout(() => {
      onClose(arg0, arg1)
      setExiting(false)
    }, timeout)
  }

  const headingId = `bpk-bottom-sheet-heading-${id}`;
  const dialogClassName = getClassName(
    'bpk-bottom-sheet',
    wide && 'bpk-bottom-sheet--wide'
    );

    return <BpkBreakpoint query={BREAKPOINTS.ABOVE_MOBILE}>
      {(isAboveMobile: boolean) =>
      <BpkDialogWrapper
      ariaLabelledby={ariaLabelledby}
      dialogClassName={dialogClassName}
      id={id}
      isOpen={isOpen}
      onClose={(
        arg0?: TouchEvent | MouseEvent | KeyboardEvent | SyntheticEvent<HTMLDialogElement, Event>,
        arg1?: {
          source: 'ESCAPE' | 'DOCUMENT_CLICK';
        }) => handleClose( isAboveMobile ? 0 : animationTimeout, arg0, arg1)}
      exiting={exiting}
      transitionClassNames={{
        appear: getClassName('bpk-bottom-sheet--appear'),
        appearActive: getClassName('bpk-bottom-sheet--appear-active'),
        exit: getClassName('bpk-bottom-sheet--exit')
      }}
      closeOnEscPressed={closeOnEscPressed}
      closeOnScrimClick={closeOnScrimClick}
      timeout={{appear: animationTimeout, exit: isAboveMobile ? 0 : animationTimeout}}
      >
      <>
        <header className={getClassName('bpk-bottom-sheet--header')}>
          <BpkNavigationBar
            id={headingId}
            title={title &&
              <BpkText id={headingId} textStyle={TEXT_STYLES.label1} tagName="h2">{title}</BpkText>
            }
            leadingButton={
              <BpkCloseButton
                label={closeLabel}
                onClick={(
                  arg0?: TouchEvent | MouseEvent | KeyboardEvent | SyntheticEvent<HTMLDialogElement, Event>,
                  arg1?: {
                    source: 'ESCAPE' | 'DOCUMENT_CLICK';
                  }) => handleClose( isAboveMobile ? 0 : animationTimeout, arg0, arg1)}
              />
            }
            trailingButton={
              actionText && onAction ? (
                <BpkButtonLink
                  onClick={onAction}
                >
                  {actionText}
                </BpkButtonLink>
              ) :
              null
            }
          />
        </header>
        <div className={getClassName('bpk-bottom-sheet--content')}>{children}</div>
    </>
    </BpkDialogWrapper>
  }
  </BpkBreakpoint>
}

export default BpkBottomSheet;
