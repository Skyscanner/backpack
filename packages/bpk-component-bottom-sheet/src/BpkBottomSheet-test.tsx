/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { renderToString } from 'react-dom/server';

import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import BpkBottomSheet from './BpkBottomSheet';
// mock breakpoint to always match
jest.mock('../../bpk-component-breakpoint/src/useMediaQuery', () => jest.fn(() => true));
describe('BpkBottomSheet', () => {

  const props = {
    ariaLabelledby: 'bottom-sheet',
    id: "my-bottom-sheet",
    isOpen: true,
    onClose: jest.fn(),
  }

  it('renders without crashing with all props', () => {
    expect(() => renderToString(
      <BpkBottomSheet
        actionText='Action'
        ariaLabelledby='bottom-sheet'
        closeLabel="Close"
        closeOnEscPressed
        closeOnScrimClick
        id="my-bottom-sheet"
        isOpen
        onAction={jest.fn()}
        onClose={jest.fn()}
        title="Bottom sheet title"
        wide
      >
        Bottom Sheet content
      </BpkBottomSheet>
    )).not.toThrow();
  });
  it('renders without crashing with minimum props', () => {
    expect(() => renderToString(
      <BpkBottomSheet
        ariaLabelledby='bottom-sheet'
        id="my-bottom-sheet"
        isOpen
        onClose={jest.fn()}
      >
        Bottom Sheet content
      </BpkBottomSheet>
    )).not.toThrow();
  });
  it('renders correctly with minimum prop', () => {
    const { asFragment } = render(
      <BpkBottomSheet
        ariaLabelledby='bottom-sheet'
        id="my-bottom-sheet"
        isOpen
        onClose={jest.fn()}
      >
        Bottom Sheet content
      </BpkBottomSheet>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly with wide prop', () => {
    const { asFragment } = render(
      <BpkBottomSheet
        {...props}
        wide
      >
        Bottom Sheet content
      </BpkBottomSheet>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly with action props', () => {
    const { asFragment } = render(
      <BpkBottomSheet
        {...props}
        actionText='Action'
        onAction={jest.fn()}
        wide
      >
        Bottom Sheet content
      </BpkBottomSheet>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly with ariaLabelledBy prop', () => {
    const { container } = render(
      <BpkBottomSheet
        {...props}
        ariaLabelledby='my-bottomsheet-title-id'
      >
        Bottom Sheet content
      </BpkBottomSheet>
    );

    expect(container.querySelector('dialog[aria-labelledby="my-bottomsheet-title-id"]')).toBeInTheDocument();
  });
  it('renders correctly with ariaLabel prop', () => {
    const { container } = render(
      <BpkBottomSheet
        {...props}
        ariaLabel='my a11y title'
      >
        Bottom Sheet content
      </BpkBottomSheet>

    );

    expect(container.querySelector('dialog[aria-label="my a11y title"]')).toBeInTheDocument();
  });
});
