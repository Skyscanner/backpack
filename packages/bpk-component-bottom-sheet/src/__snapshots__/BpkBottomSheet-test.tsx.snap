// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkBottomSheet renders correctly with action props 1`] = `
<DocumentFragment>
  <div
    class="bpk-dialog-wrapper"
  >
    <dialog
      aria-labelledby="bottom-sheet"
      class="bpk-dialog-wrapper--container bpk-bottom-sheet bpk-bottom-sheet--wide"
      data-open="true"
      id="my-bottom-sheet"
    >
      <div
        class="bpk-dialog-wrapper--contents"
        id="my-bottom-sheet-contents"
      >
        <header
          class="bpk-bottom-sheet--header"
        >
          <nav
            aria-labelledby="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
            class="bpk-navigation-bar bpk-navigation-bar--default"
          >
            <div
              class="bpk-navigation-bar__leading-item bpk-navigation-bar__leading-item--default"
            >
              <button
                aria-label=""
                class="bpk-close-button__default"
                title=""
                type="button"
              >
                <span
                  class="bpk-close-button-icon"
                >
                  <svg
                    aria-hidden="true"
                    height="1rem"
                    viewBox="0 0 24 24"
                    width="1rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.587 3.467a1.5 1.5 0 1 0-2.194 2.046q.036.038.074.074l6.438 6.44-6.44 6.44a1.5 1.5 0 0 0 2.122 2.12l6.44-6.438 6.44 6.44a1.5 1.5 0 0 0 2.12-2.122l-6.438-6.44 6.44-6.44a1.5 1.5 0 0 0-2.122-2.12l-6.44 6.438-6.44-6.44z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <span
              class="bpk-navigation-bar__title bpk-navigation-bar__title--default"
            >
              <span
                class="bpk-text bpk-text--label-1"
                id="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
              />
            </span>
            <div
              class="bpk-navigation-bar__trailing-item bpk-navigation-bar__trailing-item-default"
            >
              <button
                class="bpk-link"
                type="button"
              >
                Action
              </button>
            </div>
          </nav>
        </header>
        <div
          class="bpk-bottom-sheet--content"
        >
          Bottom Sheet content
        </div>
      </div>
    </dialog>
  </div>
</DocumentFragment>
`;

exports[`BpkBottomSheet renders correctly with minimum prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-dialog-wrapper"
  >
    <dialog
      aria-labelledby="bottom-sheet"
      class="bpk-dialog-wrapper--container bpk-bottom-sheet"
      data-open="true"
      id="my-bottom-sheet"
    >
      <div
        class="bpk-dialog-wrapper--contents"
        id="my-bottom-sheet-contents"
      >
        <header
          class="bpk-bottom-sheet--header"
        >
          <nav
            aria-labelledby="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
            class="bpk-navigation-bar bpk-navigation-bar--default"
          >
            <div
              class="bpk-navigation-bar__leading-item bpk-navigation-bar__leading-item--default"
            >
              <button
                aria-label=""
                class="bpk-close-button__default"
                title=""
                type="button"
              >
                <span
                  class="bpk-close-button-icon"
                >
                  <svg
                    aria-hidden="true"
                    height="1rem"
                    viewBox="0 0 24 24"
                    width="1rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.587 3.467a1.5 1.5 0 1 0-2.194 2.046q.036.038.074.074l6.438 6.44-6.44 6.44a1.5 1.5 0 0 0 2.122 2.12l6.44-6.438 6.44 6.44a1.5 1.5 0 0 0 2.12-2.122l-6.438-6.44 6.44-6.44a1.5 1.5 0 0 0-2.122-2.12l-6.44 6.438-6.44-6.44z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <span
              class="bpk-navigation-bar__title bpk-navigation-bar__title--default"
            >
              <span
                class="bpk-text bpk-text--label-1"
                id="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
              />
            </span>
          </nav>
        </header>
        <div
          class="bpk-bottom-sheet--content"
        >
          Bottom Sheet content
        </div>
      </div>
    </dialog>
  </div>
</DocumentFragment>
`;

exports[`BpkBottomSheet renders correctly with wide prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-dialog-wrapper"
  >
    <dialog
      aria-labelledby="bottom-sheet"
      class="bpk-dialog-wrapper--container bpk-bottom-sheet bpk-bottom-sheet--wide"
      data-open="true"
      id="my-bottom-sheet"
    >
      <div
        class="bpk-dialog-wrapper--contents"
        id="my-bottom-sheet-contents"
      >
        <header
          class="bpk-bottom-sheet--header"
        >
          <nav
            aria-labelledby="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
            class="bpk-navigation-bar bpk-navigation-bar--default"
          >
            <div
              class="bpk-navigation-bar__leading-item bpk-navigation-bar__leading-item--default"
            >
              <button
                aria-label=""
                class="bpk-close-button__default"
                title=""
                type="button"
              >
                <span
                  class="bpk-close-button-icon"
                >
                  <svg
                    aria-hidden="true"
                    height="1rem"
                    viewBox="0 0 24 24"
                    width="1rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.587 3.467a1.5 1.5 0 1 0-2.194 2.046q.036.038.074.074l6.438 6.44-6.44 6.44a1.5 1.5 0 0 0 2.122 2.12l6.44-6.438 6.44 6.44a1.5 1.5 0 0 0 2.12-2.122l-6.438-6.44 6.44-6.44a1.5 1.5 0 0 0-2.122-2.12l-6.44 6.438-6.44-6.44z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <span
              class="bpk-navigation-bar__title bpk-navigation-bar__title--default"
            >
              <span
                class="bpk-text bpk-text--label-1"
                id="bpk-bottom-sheet-heading-my-bottom-sheet-bpk-navigation-bar-title"
              />
            </span>
          </nav>
        </header>
        <div
          class="bpk-bottom-sheet--content"
        >
          Bottom Sheet content
        </div>
      </div>
    </dialog>
  </div>
</DocumentFragment>
`;
